sdk/tables/data-tables/src/TableBatch.ts

```diff
@@ -0,0 +1,294 @@
+
+  /**
+   * @param url Tables account url
+   * @param partitionKey partition key
+   * @param credential credential to authenticate the batch request
+   */
+  constructor(
+    url: string,
+    partitionKey: string,
+    interceptClient: TableClientLike,
+    batchGuid: string,
+    batchRequest: InnerBatchRequest,
+    credential?: TablesSharedKeyCredentialLike
+  ) {
+    this.credential = credential;
+    this.partitionKey = partitionKey;
+    this.url = url;
+    this.batchGuid = batchGuid;
+    this.batchRequest = batchRequest;
+    this.pendingOperations = [];
+
+    this.interceptClient = interceptClient;
+
+    // Depending on the auth method used we need to build the url
+    if (!credential) {
+      // When authenticating with SAS we need to add the SAS token after $batch
+      const urlParts = url.split("?");
+      this.url = urlParts[0];
+      const sas = urlParts.length > 1 ? `?${urlParts[1]}` : "";
+      this.url = `${this.url}/$batch${sas}`;
+    } else {
+      // When using a SharedKey credential no SAS token is needed
+      this.url = `${this.url}/$batch`;
+    }
+  }
+
+  /**
+   * Adds a createEntity operation to the batch
+   * @param entity Entity to create
+   */
+  public createEntity<T extends object>(entity: TableEntity<T>): void {
+    this.checkPartitionKey(entity.partitionKey);
+    this.pendingOperations.push(this.interceptClient.createEntity(entity));
+  }
+
+  /**
+   * Adds a createEntity operation to the batch per each entity in the entities array
+   * @param entitites Array of entities to create
+   */
+  public createEntities<T extends object>(entitites: TableEntity<T>[]): void {
+    for (const entity of entitites) {
+      this.checkPartitionKey(entity.partitionKey);
+      this.pendingOperations.push(this.interceptClient.createEntity(entity));
+    }
+  }
+
+  /**
+   * Adds a deleteEntity operation to the batch
+   * @param partitionKey partition key of the entity to delete
+   * @param rowKey row key of the entity to delete
+   * @param options options for the delete operation
+   */
+  public deleteEntity(
+    partitionKey: string,
+    rowKey: string,
+    options?: DeleteTableEntityOptions
+  ): void {
+    this.checkPartitionKey(partitionKey);
+    this.pendingOperations.push(this.interceptClient.deleteEntity(partitionKey, rowKey, options));
+  }
+
+  /**
+   * Adds an updateEntity operation to the batch
+   * @param entity entity to update
+   * @param mode update mode (Merge or Replace)
+   * @param options options for the update operation
+   */
+  public updateEntity<T extends object>(
+    entity: TableEntity<T>,
+    mode: UpdateMode,
+    options?: UpdateTableEntityOptions
+  ): void {
+    this.checkPartitionKey(entity.partitionKey);
+    this.pendingOperations.push(this.interceptClient.updateEntity(entity, mode, options));
+  }
+
+  /**
+   * Submits the operations in the batch
+   */
+  public async submitBatch(): Promise<any> {
+    await Promise.all(this.pendingOperations);
```

Comment:
Is the order in which `Promise.all()` resolves the promises specified?  It was not obvious to me from reading MDN that this was the case.  Assuming that the resolution order is unspecified, should we instead just do an explicit loop here and await in order?  I'm slightly worried about the batch being built "out of order" and maybe that leading to something observable, although all the cases I can think of to observe this would involve things like "creating and deleting an entity as part of a batch".

Reply:
Awaiting in order doesn't guarantee order of execution once they've been kicked off in parallel. It would only check if they were settled or not in order.

I think this should be safe, as you can't have conflicting operations within a single batch regardless of order, right?

Reply:
Would it be safer to store an array of callbacks instead of promises, and then call/await each of them on submit? I wonder if we can run into race-conditions and get a malformed body

Reply:
Hmm malformed in what way? No promise code could run simultaneously, so unless the code did something like 

```
// modify the body in a way that makes it incorrect
await something();
// modify the body in a way that makes it correct again, hope nobody else touched it
```

I think we should be okay?

Reply:
Yeah it looks like the policy calls `appendSubRequestToBody` which synchronously adds something to the body array, and since we don't call `getHttpRequestBody` until all promises are settled (e.g. all body parts are added to the array), we'll have them all.

Reply:
Since we are pushing to the array and not mutating the body, the worst that can happen is to have the operations added to the body in a different order as Matt mentioned.

I just checked and the service doesn't seem to allow multiple changes to the same rowKey within a batch so it looks like this probably wouldn't be an issue.

I think I'll keep the current approach as it has the extra benefit of computing start computing the body in the background as soon as the batch operations are called, so submitBatch will be faster than if we await each callback during submitBatch.

Does this make sense?


This is the error that the service gives us back

```javascript
{
  status: 202,
  subResponses: [
    {
      status: 400,
      body: {
        "odata.error": {
          code: "InvalidDuplicateRow",
          message: {
            lang: "en-US",
            value:
              "1:The batch request contains multiple changes with same row key. An entity can appear only once in a batch request. RequestId:456aeab3-b002-0014-0bef-986a9d000000 Time:2020-10-02T19:11:31.6622404Z",
          }
        }
      }
    }
  ]
}
```

Reply:
Yeah, that makes sense to me.  Thanks for checking to see that the service rejects such cases.